swap/swap_test.go

```diff
@@ -682,3 +686,246 @@ func testDeploy(ctx context.Context, backend cswap.Backend, swap *Swap) (err err
+	defer os.RemoveAll(dir)
+
+	// cheque with an invalid signature
+	testCheque := newTestCheque()
+	testCheque.Sig = manipulateSignature(testChequeSig)
+	if err := peer.verifyChequeProperties(testCheque); err == nil {
+		t.Fatalf("accepted cheque with invalid signature")
+	}
+
+	// cheque with wrong contract
+	testCheque = newTestCheque()
+	testCheque.Contract = beneficiaryAddress
+	testCheque.Sig, _ = swap.signContentWithKey(testCheque, ownerKey)
+	if err := peer.verifyChequeProperties(testCheque); err == nil {
+		t.Fatalf("accepted cheque with wrong contract")
+	}
+
+	// cheque with wrong beneficiary
+	testCheque = newTestCheque()
+	testCheque.Beneficiary = ownerAddress
+	testCheque.Sig, _ = swap.signContentWithKey(testCheque, ownerKey)
+	if err := peer.verifyChequeProperties(testCheque); err == nil {
+		t.Fatalf("accepted cheque with wrong beneficiary")
+	}
+
+	// cheque with non-zero timeout
+	testCheque = newTestCheque()
+	testCheque.Timeout = 10
+	testCheque.Sig, _ = swap.signContentWithKey(testCheque, ownerKey)
+	if err := peer.verifyChequeProperties(testCheque); err == nil {
+		t.Fatalf("accepted cheque with non-zero timeout")
+	}
+}
+
+// TestPeerVerifyChequeAgainstLast tests that verifyChequeAgainstLast accepts a cheque with higher serial and amount
+func TestPeerVerifyChequeAgainstLast(t *testing.T) {
+	_, peer, dir := newTestSwapAndPeer(t)
+	defer os.RemoveAll(dir)
+
+	increase := uint64(10)
+	oldCheque := newTestCheque()
+	newCheque := newTestCheque()
+
+	newCheque.Serial = oldCheque.Serial + 1
+	newCheque.Amount = oldCheque.Amount + increase
+
+	if err := peer.verifyChequeAgainstLast(newCheque, oldCheque, increase); err != nil {
+		t.Fatalf("failed to verify cheque compared to old cheque: %s", err.Error())
+	}
+}
+
+// TestPeerVerifyChequeAgainstLastInvalid tests that verifyChequeAgainstLast rejects cheques with lower serial or amount or an unexpected value
+func TestPeerVerifyChequeAgainstLastInvalid(t *testing.T) {
+	_, peer, dir := newTestSwapAndPeer(t)
+	defer os.RemoveAll(dir)
+
+	increase := uint64(10)
+
+	// cheque with higher amount but same serial
+	oldCheque := newTestCheque()
+	newCheque := newTestCheque()
+	newCheque.Amount = oldCheque.Amount + increase
+
+	if err := peer.verifyChequeAgainstLast(newCheque, oldCheque, increase); err == nil {
+		t.Fatal("accepted a cheque with same serial")
+	}
+
+	// cheque with higher serial but same amount
+	oldCheque = newTestCheque()
+	newCheque = newTestCheque()
+	newCheque.Serial = oldCheque.Serial + 1
+
+	if err := peer.verifyChequeAgainstLast(newCheque, oldCheque, increase); err == nil {
+		t.Fatal("accepted a cheque with same amount")
+	}
+
+	// cheque with amount != increase
+	oldCheque = newTestCheque()
+	newCheque = newTestCheque()
+	newCheque.Serial = oldCheque.Serial + 1
+	newCheque.Amount = oldCheque.Amount + increase + 5
+
+	if err := peer.verifyChequeAgainstLast(newCheque, oldCheque, increase); err == nil {
+		t.Fatal("accepted a cheque with unexpected amount")
+	}
```

Comment:
Great tests, do we need to add one if there is actually no existing cheque (first one) or is this covered?

Reply:
This particular test tests the `verifyChequeAgainstLast` function which is only ever called if there was a prior cheque.

What is missing is a test for `processAndVerifyCheque` with an invalid cheque if it is the first cheque. I will add one.